extensions/interactions/CodeRepl/CodeRepl.js

```diff
@@ -146,3 +146,156 @@ oppia.directive('oppiaShortResponseCodeRepl', [
     };
   }
 ]);
+
+oppia.factory('codeReplNormalizationService', [function() {
+  var removeLeadingWhitespace = function(str) {
+    return str.replace(/^\s+/g, '');
+  };
+  var removeTrailingWhitespace = function(str) {
+    return str.replace(/\s+$/g, '');
+  };
+  return {
+    getNormalizedCode: function(codeString) {
+      /*
+       * Normalizes a code string (which is assumed not to contain tab
+       * characters). In particular:
+       *
+       * - Strips out lines that start with '#' (comments), possibly preceded by
+       *     whitespace.
+       * - Trims trailing whitespace on each line.
+       * - Removes blank newlines.
+       * - Make the indentation level four spaces.
+       */
+      // TODO(sll): Augment this function to strip out comments that occur at the
+      // end of a line. However, be careful with lines where '#' is contained in
+      // quotes or the character is escaped.
+      var FOUR_SPACES = '    ';
+      // Maps the number of spaces at the beginning of a line to an int specifying
+      // the desired indentation level.
+      var numSpaceToDesiredIndentLevel = {
+          0: 0,
+      };
+
+      var codeLines = removeTrailingWhitespace(codeString).split('\n');
+      var normalizedCodeLines = [];
+      codeLines.forEach(function(line) {
+        if (removeLeadingWhitespace(line).startsWith('#')) {
+          return;
+        }
+        line = removeTrailingWhitespace(line);
+        if (!line) {
+          return;
+        }
+
+        var numSpaces = line.length - removeLeadingWhitespace(line).length;
+
+        var maxNumSpaces = Object.keys(numSpaceToDesiredIndentLevel).reduce(
+          function(max, key) {
+            return Math.max(key, max);
+          }, -Infinity);
+        if (numSpaces > maxNumSpaces) {
+          // Add a new indentation level
+          numSpaceToDesiredIndentLevel[numSpaces] =
+            Object.keys(numSpaceToDesiredIndentLevel).length;
```

Comment:
Feels like you want to break out a new variable above to avoid recomputing this here:

    var existingIndentLevels = Object.keys(numSpaceToDesiredIndentLevel);